sound/soc/sof/sof-audio.c

```diff
@@ -256,29 +256,125 @@ static int sof_setup_pipeline_connections(struct snd_sof_dev *sdev,
 	return 0;
 }
 
+/*
+ * free all widgets in the sink path starting from the source widget
+ * (DAI type for capture, AIF type for playback)
+ */
+static int sof_free_widgets_in_path(struct snd_sof_dev *sdev, struct snd_soc_dapm_widget *widget,
+				    int dir)
+{
+	struct snd_soc_dapm_path *p;
+	int ret;
+	int ret1 = 0;
+
+	/* free all widgets even in case of error to keep use counts balanced */
+	snd_soc_dapm_widget_for_each_sink_path(widget, p) {
+		if (p->sink->dobj.private && widget->dobj.private) {
+			if (WIDGET_IS_AIF(widget->id) || WIDGET_IS_DAI(widget->id)) {
+				ret = sof_widget_free(sdev, widget->dobj.private);
+				if (ret < 0)
+					ret1 = ret;
+			}
+
+			ret = sof_widget_free(sdev, p->sink->dobj.private);
+			if (ret < 0)
+				ret1 = ret;
+
+			ret = sof_free_widgets_in_path(sdev, p->sink, dir);
+			if (ret < 0)
+				ret1 = ret;
+		}
+	}
+
+	return ret1;
+}
+
+/*
+ * set up all widgets in the sink path starting from the source widget
+ * (DAI type for capture, AIF type for playback).
+ * The error path in this function ensures that all successfully set up widgets getting freed.
+ */
+static int sof_set_up_widgets_in_path(struct snd_sof_dev *sdev, struct snd_soc_dapm_widget *widget,
+				      int dir)
+{
+	struct snd_soc_dapm_path *p;
+	int ret;
+
+	snd_soc_dapm_widget_for_each_sink_path(widget, p) {
+		if (p->sink->dobj.private && widget->dobj.private) {
+			if (WIDGET_IS_AIF(widget->id) || WIDGET_IS_DAI(widget->id)) {
+				ret = sof_widget_setup(sdev, widget->dobj.private);
+				if (ret < 0)
+					return ret;
+			}
+
+			ret = sof_widget_setup(sdev, p->sink->dobj.private);
+			if (ret < 0) {
+				sof_widget_free(sdev, widget->dobj.private);
+				return ret;
+			}
+
+			ret = sof_set_up_widgets_in_path(sdev, p->sink, dir);
+			if (ret < 0) {
+				if (WIDGET_IS_AIF(widget->id) || WIDGET_IS_DAI(widget->id))
+					sof_widget_free(sdev, widget->dobj.private);
+				sof_widget_free(sdev, p->sink->dobj.private);
+				return ret;
+			}
+		}
+	}
+
+	return 0;
+}
+
+static int
+sof_setup_or_free_widgets_in_order(struct snd_sof_dev *sdev, struct snd_soc_dapm_widget_list *list,
+				   bool dir, bool setup)
+{
+	struct snd_soc_dapm_widget *widget;
+	int ret, i;
+
+	for_each_dapm_widgets(list, i, widget) {
+		/* starting widget for playback is AIF type */
+		if (dir == SNDRV_PCM_STREAM_PLAYBACK && !WIDGET_IS_AIF(widget->id))
+			continue;
+
+		/* starting widget for capture is DAI type */
+		if (dir == SNDRV_PCM_STREAM_CAPTURE && !WIDGET_IS_DAI(widget->id))
+			continue;
```

Comment:
You are looking up the start of the path, right? Is it expected that you will have more than one start entry?
```c
struct snd_soc_dapm_widget *widget;
bool found = false;
int i;

for_each_dapm_widgets(list, i, widget) {
	if (dir == SNDRV_PCM_STREAM_PLAYBACK && WIDGET_IS_AIF(widget->id) ||
	    dir == SNDRV_PCM_STREAM_CAPTURE && WIDGET_IS_DAI(widget->id)) {
		found = true;
		break;
	}
}

if (found) {
	int ret;

	if (setup)
		ret = sof_set_up_widgets_in_path(sdev, widget, dir);
	else
		ret = sof_free_widgets_in_path(sdev, widget, dir);

	if (ret < 0) {
		dev_err(sdev->dev, "Failed to %s connected widgets\n",
			setup ? "set up" : "free");
		return ret;
	}
}

return 0;
```
Is it considered OK if we do not find the start?

Reply:
Yes, there can be more than one starting widget. I cannot quite envision this case for playback but for capture it is certainly a possibility.

Reply:
beep tone played on multiple outputs...